File: cses/Transfer Speeds Sum.cpp
#include <bits/stdc++.h>
 
using namespace std;
 
struct UF {
	vector<int> parent, rank, len;
 
	UF(int n) {
		parent.resize(n);
		rank.assign(n, 0);
		len.assign(n, 1);
 
		iota(parent.begin(), parent.end(), 0);
	}
 
	int find_set(int u) {
		return u == parent[u] ? u : parent[u] = find_set(parent[u]);
	}
 
	void join(int u, int v) {
		if(rank[u] > rank[v])
			swap(u, v);
 
		parent[u] = v;
		len[v] += len[u];
		rank[v] += rank[u] == rank[v]; 
	}
};
 
int main() {
 
	ios_base::sync_with_stdio(false);
	cin.tie(0);
 
	int n;
	vector<tuple<int, int, int>> edges;
	int64_t sum {};
 
	cin >> n;
 
	UF uf(n);
	
	for(int i = 1; i < n; ++i) {
		int a, b, x;
 
		cin >> a >> b >> x;
 
		edges.emplace_back(x, a - 1, b - 1);
	}
 
	sort(edges.rbegin(), edges.rend());
 
	for(auto [x, a, b] : edges) {
		a = uf.find_set(a);
		b = uf.find_set(b);
 
		sum += 1LL * uf.len[a] * uf.len[b] * x;
 
		uf.join(a, b);
	}
 
	cout << sum << '\n';
 
	return 0;
}

File: cses/Transfer Speeds Sum.rs
#![allow(unused_imports)]
use std::collections::{VecDeque};
use std::io::{self, Error, ErrorKind, Write};

use crate::data_structures::UnionFind;

struct Scanner {
   buffer: VecDeque<String>
}

impl Scanner {
 
   fn new() -> Self {
      Self {
         buffer: VecDeque::new()
      }
   }
   
   fn next<T: std::str::FromStr>(&mut self) -> io::Result<T> {
 
      if self.buffer.is_empty() {
         let mut input = String::new();
         
         match std::io::stdin().read_line(&mut input) {
            Ok(0) => {
               return Err(Error::new(ErrorKind::UnexpectedEof, "End Of Line"));
            } Ok(_) => {
               
            } Err(e) => {
               return Err(e);
            }
         }
 
         self.buffer = input.split_whitespace()
                            .map(|x| x.to_string())
                            .collect();
      }
 
      let front = self.buffer.pop_front().unwrap();
      
      Ok(front.parse::<T>().ok().unwrap())
   }
}

mod data_structures {

   pub struct UnionFind {
      parent: Vec<usize>,
      rank: Vec<i32>,
      size: Vec<i32>
   }

   impl UnionFind {

      pub fn new(n: usize) -> Self {

         Self {

            parent: (0..n).collect(),
            rank: vec![0; n],
            size: vec![1; n]
         }
      }

      fn find_set(&mut self, u: usize) -> usize {
         if u == self.parent[u] {
            return u;
         }

         self.parent[u] = self.find_set(self.parent[u]);

         return self.parent[u];
      }

      pub fn unite(&mut self, mut u: usize, mut v: usize, weight: i64) -> i64 {

         u = self.find_set(u);
         v = self.find_set(v);

         if self.rank[u] > self.rank[v] {
            std::mem::swap(&mut u, &mut v);
         }

         let combinations = self.size[u] as i64 * self.size[v] as i64 * weight;

         self.parent[u] = v;
         self.size[v] += self.size[u];
         self.rank[v] = if self.rank[u] == self.rank[v] {
            self.rank[v] + 1
         } else {
            self.rank[v]
         };

         combinations
      }
   }
}

fn main() {
   let stdout = std::io::stdout();
   let mut handle = std::io::BufWriter::new(stdout);
 
   let mut sc = Scanner::new();
   
   let n = sc.next::<usize>().ok().unwrap();

   let mut sum = i64::default();

   let mut uf = UnionFind::new(n);

   let mut edges = (1..n).map(|_| {
      let a = sc.next::<usize>().ok().unwrap() - 1;
      let b = sc.next::<usize>().ok().unwrap() - 1;
      let weight = sc.next::<i64>().ok().unwrap();

      (weight, a, b)
   }).collect::<Vec<_>>();

   edges.sort();
   edges.reverse();

   for &(weight, a, b) in edges.iter() {
      sum += uf.unite(a, b, weight);
   }

   writeln!(handle, "{sum}").ok();
}